pkg/s3/s3client.go

```diff
@@ -0,0 +1,167 @@
+		if aerr, ok := err.(awserr.Error); !ok {
+			return err
+		} else if aerr.Code() != s3.ErrCodeBucketAlreadyExists && aerr.Code() != s3.ErrCodeBucketAlreadyOwnedByYou {
+			return err
+		}
+	}
+	// Set lifecycle rule to purge incomplete multipart upload orphaned because of force shutdown or rescheduling or networking issue with etcd-backup-restore.
+	putBucketLifecycleConfigurationInput := &s3.PutBucketLifecycleConfigurationInput{
+		Bucket: aws.String(bucket),
+		LifecycleConfiguration: &s3.BucketLifecycleConfiguration{
+			Rules: []*s3.LifecycleRule{
+				{
+					// Note: Though as per documentation at https://docs.aws.amazon.com/AmazonS3/latest/API/API_LifecycleRule.html the Filter field is
+					// optional, if not specified the SDK API fails with `Malformed XML` error code. Cross verified same behavior with aws-cli client as well.
+					// Please do not remove it.
+					Filter: &s3.LifecycleRuleFilter{
+						Prefix: aws.String(""),
+					},
+					AbortIncompleteMultipartUpload: &s3.AbortIncompleteMultipartUpload{
+						DaysAfterInitiation: aws.Int64(7),
+					},
+					Status: aws.String(s3.ExpirationStatusEnabled),
+				},
+			},
+		},
+	}
+
+	_, err := c.S3.PutBucketLifecycleConfigurationWithContext(ctx, putBucketLifecycleConfigurationInput)
+	return err
+}
+
+// DeleteBucketIfExists deletes the s3 bucket with name <bucket>. If it does not exist,
+// no error is returned.
+func (c *S3Client) DeleteBucketIfExists(ctx context.Context, bucket string) error {
+	if _, err := c.S3.DeleteBucketWithContext(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)}); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return nil
+			}
+			if aerr.Code() == "BucketNotEmpty" {
+				if err := c.DeleteObjectsWithPrefix(ctx, bucket, ""); err != nil {
+					return err
+				}
+				return c.DeleteBucketIfExists(ctx, bucket)
+			}
+		}
+		return err
+	}
+	return nil
+}
+
+// DeleteObjectsWithPrefix deletes the s3 objects with the specific <prefix> from <bucket>. If it does not exist,
+// no error is returned.
+func (c *S3Client) DeleteObjectsWithPrefix(ctx context.Context, bucket, prefix string) error {
+	in := &s3.ListObjectsInput{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	}
+
+	var delErr error
+	if err := c.S3.ListObjectsPagesWithContext(ctx, in, func(page *s3.ListObjectsOutput, lastPage bool) bool {
+		objectIDs := make([]*s3.ObjectIdentifier, 0)
+		for _, key := range page.Contents {
+			obj := &s3.ObjectIdentifier{
+				Key: key.Key,
+			}
+			objectIDs = append(objectIDs, obj)
+		}
+
+		if len(objectIDs) != 0 {
+			if _, delErr = c.S3.DeleteObjectsWithContext(ctx, &s3.DeleteObjectsInput{
+				Bucket: aws.String(bucket),
+				Delete: &s3.Delete{
+					Objects: objectIDs,
+					Quiet:   aws.Bool(true),
+				},
+			}); delErr != nil {
+				return false
+			}
+		}
+		return !lastPage
+	}); err != nil {
+		return err
```

Comment:
Let's wrap all errors coming from external packages.
```suggestion
		return errors.Wrap(err, "could not list objects")
```